doc/api/quic.md

```diff
@@ -0,0 +1,2276 @@
+  `QuicSession` enters the `Ready` state, it may be used to exchange
+  application data using `QuicStream` instances.
+* `Closed` - Entere as soon as the `QuicSession` connection has been
+  terminated.
+
+New instances of `QuicClientSession` are created using the `connect()`
+function on a `QuicSocket` as in the example below:
+
+```js
+const { createQuicSocket } = require('net');
+
+// Create a QuicSocket associated with localhost and port 1234
+const socket = createQuicSocket({ endpoint: { port: 1234 } });
+
+const client = socket.connect({
+  address: 'example.com',
+  port: 4567,
+  alpn: 'foo'
+});
+```
+
+As soon as the `QuicClientSession` is created, the `address` provided in
+the connect options will be resolved to an IP address (if necessary), and
+the TLS 1.3 handshake will begin. The `QuicClientSession` cannot be used
+to exchange application data until after the `'secure'` event has been
+emitted by the `QuicClientSession` object, signaling the completion of
+the TLS 1.3 handshake.
+
+```js
+client.on('secure', () => {
+  // The QuicClientSession can now be used for application data
+});
+```
+
+New instances of `QuicServerSession` are created internally by the
+`QuicSocket` if it has been configured to listen for new connections
+using the `listen()` method.
+
+```js
+const key = getTLSKeySomehow();
+const cert = getTLSCertSomehow();
+
+socket.listen({
+  key,
+  cert,
+  alpn: 'foo'
+});
+
+socket.on('session', (session) => {
+  session.on('secure', () => {
+    // The QuicServerSession can now be used for application data
+  });
+});
+```
+
+As with client `QuicSession` instances, the `QuicServerSession` cannot be
+used to exhange application data until the `'secure'` event has been emitted.
+
+### QuicSession and ALPN
+
+QUIC uses the TLS 1.3 [ALPN][] ("Application-Layer Protocol Negotiation")
+extension to identify the application level protocol that is using the QUIC
+connection.  Every `QuicSession` instance has an ALPN identifier that *must* be
+specified in either the `connect()` or `listen()` options. ALPN identifiers that
+are known to Node.js (such as the ALPN identifier for HTTP/3) will alter how the
+`QuicSession` and `QuicStream` objects operate internally, but the QUIC
+implementation for Node.js has been designed to allow any ALPN to be specified
+and used.
+
+### QuicStream
+
+Once a `QuicSession` transitions to the `Ready` state, `QuicStream` instances
+may be created and used to exchange application data. On a general level, all
+`QuicStream` instances are simply Node.js Duplex Streams that allow
+bidirectional data flow between the QUIC client and server. However, the
+application protocol negotiated for the `QuicSession` may alter the semantics
+and operation of a `QuicStream` associated with the session. Specifically,
+some features of the `QuicStream` (e.g. headers) are enabled only if the
+application protocol selected is known by Node.js to support those features.
```

Comment:
Yes and no. In the current implementation, the built in header support is implemented at the native/C++ level and is set up such that any specialization of `QuicApplication` can implement support for headers. Anyone implementing a user-defined alpn at the JavaScript level, however, will have to implement their own idea of headers. This is because QUIC itself has absolutely no concept of application headers. HTTP/3 implements header support using additional special purpose unidirectional streams and stream-level encoded data that gets rather complicated. 

If grpc ran over http/3, then it would leverage the HTTP/3 support here and use the header mechanism provided already. If grpc evolves into it's own QUIC application protocol, it would need to develop it's own notion of how headers are supported. This is not because of any design decision in the QUIC implementation ... QUIC itself has no concept of headers that we can build a generic mechanism on top of.

Reply:
That makes sense. I think the docs should be bit more clear on that, basically the info that the headers aren't a QUIC thing, but http/3 specific, so no one comes asking for how to use it for their protocol (like I just did :-).